src/jobs/concretes/block.job.ts

```diff
@@ -0,0 +1,120 @@
+            logger.info("[BlockJob] Starting cycle.");
+
+            try {
+                const proposedBlock: Block = await this.generateProposedBlockAsync();
+                // TODO: Shouldn't the peer methods be async?
+               // this.peer.broadcastMessage(MessageType.NEW_PROPOSED_BLOCK, JSON.stringify(proposedBlock));
+
+                const states: State[] = await this.dataAccessLayer.getStatesAsync();
+                const victoriousBlock: Block = await this.lotteryService
+                                                            .drawWinningBlock(proposedBlock.blockHeader.parentHash,
+                                                                              states);
+                await this.transactionService.updatePendingTransactions(victoriousBlock);
+                // TODO: Shouldn't the peer methods be async?
+                // this.peer.broadcastMessage(MessageType.NEW_VICTORIOUS_BLOCK, JSON.stringify(victoriousBlock));
+
+                resolve();
+            } catch (error) {
+                logger.error(error);
+            
+                reject(error);
+                return;
+            }
+        });
+    }
+
+    private async generateProposedBlockAsync(): Promise<Block> {
+        return new Promise(async (resolve, reject) => {
+            const blockChain: Block[] = await this.dataAccessLayer.getBlockchainAsync();
+
+            if (blockChain.length === 0) {
+                reject(new BlockJobException("[BlockJob] Skipped current cycle because of empty blockchain."));
+                return;
+            }
+
+            if (!this.keyPair) {
+                reject(new BlockJobException("[BlockJob] Skipped current cycle because the keypair is undefined."));
+                return;
+            }
+            
+            const lastBlock: Block = blockChain[blockChain.length - 1];
+            const proposedBlock: Block = await this.proposedBlockGenerator
+                                                        .generateProposedBlockAsync(
+                                                            lastBlock,
+                                                            this.queueStore.pendingTransactionQueue,
+                                                            this.constantStore.VALIDATOR_VERSION,
+                                                            this.keyPair.publicKey);
+            
+            resolve(proposedBlock);
+        });
+    }
+
+    private async getOrGenerateKeyPairAsync(): Promise<{ publicKey: string; privateKey: string; }> {
+        return new Promise(async (resolve) => {
+            logger.info("[BlockJob] Grabbing and/or generating keypair.");
+
+            if (await this.fileUtils.fileExistsAsync(this.constantStore.KEYS_FILE_PATH)) {
+                resolve(await this.fileUtils.readFileAsync(this.constantStore.KEYS_FILE_PATH));
+            }
+
+            // TODO: Get keypair -> save key? 
```

Comment:
Implement or remove this TODO

Reply:
This is an intermediate in progress Pull Request, this code should not be merged. This means that there are still TODOs left, otherwise it would not be an intermediate in-progress PR. This comment obviously applies to all your TODO notes 👍 .